internal/controller/apiextensions/claim/configurator.go

```diff
@@ -134,13 +122,18 @@ func (c *APIDryRunCompositeConfigurator) Configure(ctx context.Context, cm resou
 	ucp.SetClaimReference(proposed)
 
 	if !meta.WasCreated(cp) {
-		// The API server returns an available name derived from
-		// generateName when we perform a dry-run create. This name is
-		// likely (but not guaranteed) to be available when we create
-		// the composite resource. If the API server generates a name
-		// that is unavailable it will return a 500 ServerTimeout error.
-		cp.SetGenerateName(fmt.Sprintf("%s-", cm.GetName()))
-		return errors.Wrap(c.client.Create(ctx, cp, client.DryRunAll), errName)
+		// UUID can have several textual representations
+		// https://en.wikipedia.org/wiki/Universally_unique_identifier#Textual_representation
+		// For robustness, try to parse it first
+		id, err := uuid.Parse(string(cm.GetUID()))
+		if err != nil {
+			return err
+		}
+		// Construct XR name following the convention <claim_name>-<claim_uid_last_64_bits>
+		// risk for collision is super low (2^-12 for 10k claims with the same name)
+		// https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions
+		// therefore we do not add any logic that would handle it.
+		cp.SetName(fmt.Sprintf("%s-%s", cm.GetName(), hex.EncodeToString(id[8:])))
```

Comment:
Hex encoding means 4 bits per character. Hence, we are appending a 64/4=16 long random string, plus the `-`. I.e. we are losing 17 characters out of 63 allowed. Before this change, with Kube doing the name generation we only lost 6 characters. This breaks claims with long names (> 46 chars; before 57 were fine).

In other words: this change is a breaking API change for Crossplane.

Reply:
There are libraries for base62 encoding (2*26+ 10). With that we nearly squeeze 6 bits into one character, leading to 11ish characters for the suffix.

Assuming we use a 5 char suffix and base62, what would that mean for the math?

Reply:
This made my day: https://x.com/the_sttts/status/1711702505590063478?s=20

Reply:
Actually, names must be lower-case I guess. So we have only base36 at best.

Reply:
> Actually, names must be lower-case I guess. So we have only base36 at best.

Yes, and with base36 we save just 3 chars.

If we wanna take just 48bits instead of 64, then with 1000 claims, the probability for collision is 10<sup>-7</sup> and then using base36 encoding can get us to 10 char suffix.

> In other words: this change is a breaking API change for Crossplane.

It is already reported in https://github.com/crossplane/crossplane/issues/4645, not sure if we advertise, nor check name size limitations right now. Moving to annotations is going to solve this problem.

@sttts let me know if we should go with base36 encoding here.

Reply:
How hard is it to optimistically choose a suffix in kube style (maybe even using their helper), go with it, and in case the XR exists but belongs to some other claim we retry by choosing another suffix and store it in the claim ?

Reply:
Still a breaking change.

Reply:
we do check not report back to use what the max claim name length.